File: web/src/model/registry/elements/instructionDisplay.ts
/**
 * The instruction display (InstructionDisplayElm.java, XML type "ins", port
 * dump 434): a bus of `busWidth` inputs that forms an integer from the per-bit
 * logic levels (a post above `threshold` sets its bit), then maps that integer
 * through a lookup table to a drawn string. Pure readout: it adds no matrix
 * unknown (getVoltageSourceCount() == 0).
 *
 * Upstream's text dump code is effectively 0 (it never overrides
 * getDumpType), so the port uses 434, the smallest unused code in the modern
 * XML-era block. The lookup table rides in `e.text`, written as one escaped
 * token. See feature/instruction-display.md for the geometry deviation
 * (upstream stacks every bus post on one (x,y) point and relies on the post
 * index for node identity; the port spreads them vertically so each bit is a
 * distinct node).
 */

import { canvasFont, limbColor, line } from '../../../render/draw';
import type { CircuitElement, DrawContext, ElementDef, Point } from '../../types';

/** The default lookup table upstream seeds the element with. */
export const DEFAULT_LOOKUP = '0=text0\n1=text1\n0x2-0xF=other ({a})\n';

/** The bus width, clamped like the engine: truncated and capped to 1..32. */
export function instructionDisplayBits(value: number): number {
  if (!Number.isFinite(value) || value < 1) return 1;
  if (value > 32) return 32;
  return Math.round(value);
}

function busWidth(e: CircuitElement): number {
  return instructionDisplayBits(e.params.busWidth ?? 4);
}

/** The integer formed from the bus posts (InstructionDisplayElm.readInputValue). */
export function instructionDisplayValue(
  voltages: (number | undefined)[],
  busWidth: number,
  threshold: number,
): number {
  let value = 0;
  for (let i = 0; i < busWidth; i++) {
    if ((voltages[i] ?? 0) > threshold) value |= 1 << i;
  }
  return value;
}

/** Parses one lookup key, handling decimal and 0x/0b prefixes. */
function parseKey(s: string): number {
  const t = s.trim();
  const prefixed = t.startsWith('0x') || t.startsWith('0X') || t.startsWith('0b') || t.startsWith('0B');
  if (prefixed) {
    const rest = t.slice(2).trim();
    const radix = t[1] === 'b' || t[1] === 'B' ? 2 : 16;
    const n = parseInt(rest, radix);
    return Number.isFinite(n) ? n : 0;
  }
  const n = parseInt(t, 10);
  return Number.isFinite(n) ? n : 0;
}

interface LookupEntry {
  lo: number;
  hi: number;
  template: string;
}

/** Splits the multi-line lookup text into rows; a line without `=` is skipped. */
function parseLookup(lookup: string): LookupEntry[] {
  const entries: LookupEntry[] = [];
  for (const rawLine of lookup.split('\n')) {
    const line = rawLine.trim();
    if (line.length === 0) continue;
    const eq = line.indexOf('=');
    if (eq < 0) continue;
    const key = line.slice(0, eq).trim();
    const template = line.slice(eq + 1);
    const start = key.startsWith('0x') || key.startsWith('0X') || key.startsWith('0b') || key.startsWith('0B') ? 2 : 0;
    const dash = key.slice(start).indexOf('-');
    if (dash >= 0) {
      const d = dash + start;
      entries.push({ lo: parseKey(key.slice(0, d)), hi: parseKey(key.slice(d + 1)), template });
    } else {
      const k = parseKey(key);
      entries.push({ lo: k, hi: k, template });
    }
  }
  return entries;
}

/** Renders a template, substituting `{expr}` with the result (variable `a`). */
function renderTemplate(template: string, value: number): string {
  let out = '';
  let pos = 0;
  while (pos < template.length) {
    const open = template.indexOf('{', pos);
    if (open < 0) {
      out += template.slice(pos);
      break;
    }
    out += template.slice(pos, open);
    const close = template.indexOf('}', open);
    if (close < 0) {
      out += template.slice(open);
      break;
    }
    const exprStr = template.slice(open + 1, close);
    try {
      // The only variable the format supports is `a`, the input value.
      // A minimal safe evaluator: numbers, `a`, + - * / and parentheses.
      const result = evalLookupExpr(exprStr, value);
      out += result;
    } catch {
      out += `{${exprStr}}`;
    }
    pos = close + 1;
  }
  return out;
}

/** A tiny arithmetic evaluator for `{a}` expressions: numbers, `a`, and
 *  `+ - * / %` with parentheses. Enough for the lookup templates. */
function evalLookupExpr(expr: string, a: number): string {
  const chars = expr.replace(/\s+/g, '');
  let i = 0;
  const parseExpr = (): number => {
    let v = parseTerm();
    while (i < chars.length && (chars[i] === '+' || chars[i] === '-')) {
      const op = chars[i++];
      const r = parseTerm();
      v = op === '+' ? v + r : v - r;
    }
    return v;
  };
  const parseTerm = (): number => {
    let v = parseFactor();
    while (i < chars.length && (chars[i] === '*' || chars[i] === '/' || chars[i] === '%')) {
      const op = chars[i++];
      const r = parseFactor();
      v = op === '*' ? v * r : op === '/' ? v / r : v % r;
    }
    return v;
  };
  const parseFactor = (): number => {
    if (chars[i] === '(') {
      i++;
      const v = parseExpr();
      if (chars[i] === ')') i++;
      return v;
    }
    if (chars[i] === 'a' || chars[i] === 'A') {
      i++;
      return a;
    }
    let num = '';
    while (i < chars.length && /[0-9.]/.test(chars[i])) num += chars[i++];
    return num.length > 0 ? parseFloat(num) : 0;
  };
  if (chars.length === 0) return `{${expr}}`;
  const v = parseExpr();
  if (i < chars.length || !Number.isFinite(v)) return `{${expr}}`;
  if (Number.isInteger(v)) return String(v);
  return String(v);
}

/**
 * Maps `value` through `lookup` to the string to draw, the port of upstream's
 * getDisplayText. No entry matches -> the decimal value.
 */
export function instructionDisplayText(value: number, lookup: string): string {
  const entries = parseLookup(lookup);
  for (const e of entries) {
    if (value >= e.lo && value <= e.hi) {
      return renderTemplate(e.template, value);
    }
  }
  return String(value);
}

/** The bus posts: a vertical stack centred on the anchor (x1, y1), one node
 *  per bit so wires connect to individual bits. */
export function instructionDisplayPosts(e: CircuitElement): Point[] {
  const n = busWidth(e);
  const grid = 16;
  const top = e.y1 - ((n - 1) * grid) / 2;
  const posts: Point[] = [];
  for (let i = 0; i < n; i++) {
    posts.push({ x: e.x1, y: Math.round(top + i * grid) });
  }
  return posts;
}

function drawInstructionDisplay(g: DrawContext, e: CircuitElement): void {
  const n = busWidth(e);
  const threshold = e.params.threshold ?? 2.5;
  const value = instructionDisplayValue(g.voltages, n, threshold);
  const text = instructionDisplayText(value, e.text ?? DEFAULT_LOOKUP);

  // A short stem from the bus anchor to the text, faint.
  line(g, { x: e.x1, y: e.y1 }, { x: e.x2, y: e.y2 }, limbColor(g, g.theme.text), 2);

  // The instruction string, centred at the text anchor (x2, y2).
  g.ctx.fillStyle = limbColor(g, g.theme.text);
  g.ctx.font = `bold ${canvasFont(14)}`;
  g.ctx.textAlign = 'center';
  g.ctx.textBaseline = 'middle';
  g.ctx.fillText(text, e.x2, e.y2);
}

export const INSTRUCTION_DISPLAY_DEF: ElementDef = {
  kind: 'instructionDisplay',
  label: 'Instruction display',
  category: 'Other',
  dumpCode: '434',
  postCount: 4, // the busWidth(4) default, for the fresh-part fallback
  postCountOf: (e) => busWidth(e),
  posts: instructionDisplayPosts,
  noDiagonal: true,
  defaultLength: 4,
  defaults: { busWidth: 4, threshold: 2.5 },
  parse: (t, e) => {
    const bw = Math.round(Number(t[0]));
    if (Number.isFinite(bw)) e.params.busWidth = instructionDisplayBits(bw);
    const th = Number(t[1]);
    if (Number.isFinite(th)) e.params.threshold = th;
    e.text = t[2] ?? DEFAULT_LOOKUP;
  },
  dump: (e) => [busWidth(e), e.params.threshold ?? 2.5, e.text ?? DEFAULT_LOOKUP],
  fields: [
    { name: 'busWidth', label: 'Bus Width', min: 1, max: 32, integer: true },
    { name: 'threshold', label: 'Threshold Voltage', unit: 'V' },
    { name: 'lookup', label: 'Lookup Table', type: 'text', target: 'text' },
  ],
  draw: drawInstructionDisplay,
};

File: engine/core/src/elements/instruction_display.rs
//! The instruction display (InstructionDisplayElm.java, XML type "ins").
//!
//! A passive readout: a bus of `busWidth` inputs forms an integer from the
//! per-bit logic levels (a post above `threshold` sets its bit), then that
//! integer is mapped through a lookup table to a string that the UI draws.
//! It contributes no matrix unknown, so it is a pure readout.
//!
//! Upstream's text dump code is effectively 0 (it never overrides
//! `getDumpType`), so the port assigns `434`. Codes 431 and 435 are also free,
//! but `434` sits in the modern 400+ XML-era block alongside the other
//! XML-era elements (relay 425/426, busSplitter 433), so it keeps the
//! assignment coherent. See `feature/instruction-display.md`.

use crate::element::{Base, Element, SimCtx};
use crate::expr::{parse as parse_expr, ExprState};
use crate::spec::ElementSpec;
use crate::stamp::Stamper;

/// The bus width, clamped like upstream's edit dialog (1..32).
fn clamp_bus_width(raw: f64) -> usize {
    let n = raw as usize;
    n.clamp(1, 32)
}

/// One parsed lookup line: a value range `[lo, hi]` and the text template.
struct LookupEntry {
    lo: i64,
    hi: i64,
    template: String,
}

/// Parses a key (`lo`, `lo-hi`, or `0x`/`0b` prefixed) into an integer.
fn parse_number(s: &str) -> i64 {
    let s = s.trim();
    let prefixed = s
        .strip_prefix("0x")
        .or_else(|| s.strip_prefix("0X"))
        .or_else(|| s.strip_prefix("0b"))
        .or_else(|| s.strip_prefix("0B"));
    if let Some(rest) = prefixed {
        // The base follows the prefix, not the suffix: "0b10" is binary, so
        // key off the prefix, not the trailing digit.
        let radix = if s.starts_with("0b") || s.starts_with("0B") {
            2
        } else {
            16
        };
        i64::from_str_radix(rest.trim(), radix).unwrap_or(0)
    } else {
        s.parse::<i64>().unwrap_or(0)
    }
}

/// Splits the multi-line lookup text into [`LookupEntry`] rows. A line with no
/// `=` is skipped; a key with a dash (past any 0x/0b prefix) is a range.
fn parse_entries(lookup: &str) -> Vec<LookupEntry> {
    let mut entries = Vec::new();
    for line in lookup.split('\n') {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let eq = match line.find('=') {
            Some(i) => i,
            None => continue,
        };
        let key = line[..eq].trim();
        let template = line[eq + 1..].to_string();
        // A dash only starts the range past the optional radix prefix.
        let start = if key.starts_with("0x")
            || key.starts_with("0X")
            || key.starts_with("0b")
            || key.starts_with("0B")
        {
            2
        } else {
            0
        };
        let dash = key[start..].find('-').map(|i| i + start);
        match dash {
            Some(d) => {
                let lo = parse_number(&key[..d]);
                let hi = parse_number(&key[d + 1..]);
                entries.push(LookupEntry { lo, hi, template });
            }
            None => {
                let k = parse_number(key);
                entries.push(LookupEntry {
                    lo: k,
                    hi: k,
                    template,
                });
            }
        }
    }
    entries
}

/// Renders one template, substituting `{expr}` with the expression result
/// evaluated with `a = value` (InstructionDisplayElm.LookupEntry.getText).
fn render_template(template: &str, value: u32) -> String {
    let mut out = String::new();
    let mut pos = 0;
    while pos < template.len() {
        let open = match template[pos..].find('{') {
            Some(i) => pos + i,
            None => {
                out.push_str(&template[pos..]);
                break;
            }
        };
        out.push_str(&template[pos..open]);
        let close = match template[open..].find('}') {
            Some(i) => open + i,
            None => {
                out.push_str(&template[open..]);
                break;
            }
        };
        let expr_str = &template[open + 1..close];
        match parse_expr(expr_str) {
            Ok(expr) => {
                let mut st = ExprState::new();
                st.values[0] = value as f64; // `a` is the input value
                let r = expr.eval(&st);
                if r.is_finite() && (r - r.floor()).abs() < 1e-9 {
                    out.push_str(&(r as i64).to_string());
                } else {
                    out.push_str(&r.to_string());
                }
            }
            Err(_) => out.push_str(&format!("{{{expr_str}}}")),
        }
        pos = close + 1;
    }
    out
}

pub struct InstructionDisplay {
    base: Base,
    bus_width: usize,
    threshold: f64,
}

impl InstructionDisplay {
    pub fn new(spec: &ElementSpec) -> Self {
        let bus_width = clamp_bus_width(spec.param("busWidth", 4.0));
        Self {
            base: Base::with_posts(bus_width),
            bus_width,
            threshold: spec.param("threshold", 2.5),
        }
    }

    /// The integer formed from the bus posts: bit `i` is set when post `i`'s
    /// voltage is above the threshold (InstructionDisplayElm.readInputValue).
    fn read_value(&self) -> u32 {
        let mut value = 0u32;
        for i in 0..self.bus_width {
            if self.base.volts.get(i).copied().unwrap_or(0.0) > self.threshold {
                value |= 1 << i;
            }
        }
        value
    }

    /// Maps `value` through `lookup`, the port of upstream's `getDisplayText`.
    pub fn display_text(value: u32, lookup: &str) -> String {
        let entries = parse_entries(lookup);
        let v = value as i64;
        for e in &entries {
            if v >= e.lo && v <= e.hi {
                return render_template(&e.template, value);
            }
        }
        value.to_string()
    }
}

impl Element for InstructionDisplay {
    fn kind(&self) -> &'static str {
        "instructionDisplay"
    }
    fn base(&self) -> &Base {
        &self.base
    }
    fn base_mut(&mut self) -> &mut Base {
        &mut self.base
    }
    fn post_count(&self) -> usize {
        self.bus_width
    }
    /// Upstream stacks all N posts on one coordinate with bit tags
    /// (`getPost(n) = new Point(x, y, n)`, InstructionDisplayElm.java:53-55);
    /// the frontend now sends them coincident, so the bit tags are what keep
    /// the inputs apart.
    fn post_bus_z(&self, post: usize) -> usize {
        post
    }
    /// Each post is an independent input; the part draws no current, so its
    /// terminals do not couple (like the meter/readout family).
    fn connects(&self, _a: usize, _b: usize) -> bool {
        false
    }
    fn calculate_current(&mut self, _ctx: &SimCtx) {
        self.base.current = 0.0;
    }
    /// A multi-post readout attributes no current to any terminal.
    fn current_into_node(&self, _post: usize) -> f64 {
        0.0
    }
    /// The readout channel reports the live bus value.
    fn value(&self) -> f64 {
        self.read_value() as f64
    }
    fn display_state(&self) -> f64 {
        self.read_value() as f64
    }
    fn set_param(&mut self, name: &str, value: f64) -> bool {
        // `busWidth` is structural (changes the post count), so it falls
        // through to a full rebuild rather than a live `set_param`.
        if name == "threshold" {
            self.threshold = value;
            true
        } else {
            false
        }
    }
    // No `stamp`/`do_step`: a readout adds no matrix unknown.
    fn stamp(&mut self, _ctx: &SimCtx, _s: &mut Stamper) {}
}
